k2/csrc/aux_labels.cc

```diff
@@ -61,4 +136,93 @@ void MapAuxLabels2(const AuxLabels &labels_in,
   start_pos.push_back(num_labels);
 }
 
+void InvertFst(const Fsa &fsa_in, const AuxLabels &labels_in, Fsa *fsa_out,
+               AuxLabels *aux_labels_out) {
+  CHECK_NOTNULL(fsa_out);
+  CHECK_NOTNULL(aux_labels_out);
+  fsa_out->arc_indexes.clear();
+  fsa_out->arcs.clear();
+  aux_labels_out->start_pos.clear();
+  aux_labels_out->labels.clear();
+
+  if (IsEmpty(fsa_in)) {
+    aux_labels_out->start_pos.push_back(0);
+    return;
+  }
+
+  auto num_states_in = fsa_in.NumStates();
+  // get the number of extra states we need to create for each state
+  // in fsa_in when inverting
+  std::vector<int32_t> expanded_states(num_states_in, 0);
+  CountExpandedStatesNum(fsa_in, labels_in, &expanded_states);
+
+  // map state in fsa_in to state in fsa_out
+  std::vector<int32_t> state_map(num_states_in, 0);
+  std::vector<int32_t> state_ids(num_states_in, 0);
+  MapStates(expanded_states, &state_map, &state_ids);
+
+  // a maximal approximation
+  int32_t num_arcs_out = labels_in.labels.size() + fsa_in.arcs.size();
+  std::vector<Arc> arcs;
+  arcs.reserve(num_arcs_out);
+  // `+1` for the end position of the last arc's olabel sequence
+  std::vector<int32_t> start_pos;
+  start_pos.reserve(num_arcs_out + 1);
+  std::vector<int32_t> labels;
+  labels.reserve(fsa_in.arcs.size());
+  int32_t final_state_in = fsa_in.FinalState();
+
+  int32_t num_non_eps_ilabel_processed = 0;
+  start_pos.push_back(0);
+  for (auto i = 0; i != fsa_in.arcs.size(); ++i) {
+    const auto &arc = fsa_in.arcs[i];
+    int32_t pos_start = labels_in.start_pos[i];
+    int32_t pos_end = labels_in.start_pos[i + 1];
+    int32_t src_state = arc.src_state;
+    int32_t dest_state = arc.dest_state;
+    if (dest_state == final_state_in) {
```

Comment:
Seems we need to create a function `IsValid` for `AuxLabels`. We need to check if `start_pos` and `labels` in this state are consistent and any arc entering final state should have an olabel with kFinalSymbol

Reply:
We'll have to decide whether the kFinalSymbol should be swapped, or if we're OK with the aux label there being 0 (and just not swap that one).  I'm not sure right now.  Lean towards 0.

Reply:
If olable is `-1 = kFinalSymbol`, we can just treat `-1` as normal symbol when inverting (remove line 183-185 above), while we always need to check if `dest_state == final_state_in` if olabel of such arc is `0` as we need to add `-1` as the ilabel of the output arc.